pkg/core/native/native_neo.go

```diff
@@ -904,6 +904,8 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
 		if !cd.Registered {
 			return errors.New("validator must be registered")
 		}
+	} else {
+		acc.LastGasPerVote = *big.NewInt(0)
```

Comment:
I think you're doing it after `PutStorageItem` now which defeats the purpose. We need a test for it.

Reply:
For this particular case we should not update last gas per vote before line 894 (`PutStorageItem`). Because we have L894 to match C#'s `GetAndChange` behaviour which always saves changed storage item even if `false` is returned in the middle of `vote` handling due to some error. It affects dumps.

What we need is to place this change exactly to the place where C# code has it, to the end of the method, right before the notification emission, so that LastGasPerVote update becomes the last change of account's state before saving. This allows to avoid bugs like c63528db fixes.

Reply:
> We need a test for it.

We need two tests, if possible: one test that checks the "successful" case of voting for null (when no error is returned from `VoteInternal`, then account's LastGasPerVote should be 0 in the end). And another one that checks that in case of early return from `VoteInternal` due to some error before L935 account's LastGasPerVote stays unchanged.